discovery/file/file.go

```diff
@@ -35,6 +38,60 @@ import (
 
 const fileSDFilepathLabel = model.MetaLabelPrefix + "filepath"
 
+// TimestampCollector is a Custom Collector for Timestamps of the files.
+type TimestampCollector struct {
+	index       int
+	filenames   []string
+	Description *prometheus.Desc
+	lock        sync.RWMutex
+
+	logger log.Logger
+}
+
+// Describe method sends the description to the channel.
+func (t *TimestampCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- t.Description
+}
+
+// SetFiles changes the filenames of the struct to the paths returned by listfiles().
+func (t *TimestampCollector) SetFiles(files []string) {
+	t.lock.Lock()
+	t.filenames = files
+	t.lock.Unlock()
+}
+
+// Collect creates constant metrics for each file with last modified time of the file.
+func (t *TimestampCollector) Collect(ch chan<- prometheus.Metric) {
+	t.lock.RLock()
+	files := t.filenames
+	t.lock.RUnlock()
+	for i := 0; i < len(files); i++ {
+		info, err := os.Stat(files[i])
+		if err != nil {
+			level.Error(t.logger).Log("msg", "Error getting fileinfo", "filename", files[i], "err", err)
+			continue
+		}
+		ch <- prometheus.MustNewConstMetric(
+			t.Description,
+			prometheus.GaugeValue,
+			float64(info.ModTime().Unix()),
+			files[i],
+		)
+	}
+}
+
+// NewTimestampCollector creates a TimestampCollector.
+func NewTimestampCollector(i int) *TimestampCollector {
+	return &TimestampCollector{
+		Description: prometheus.NewDesc(
+			"prometheus_sd_file_timestamp",
+			"Timestamp of files read by FileSD",
+			[]string{"filename"},
+			prometheus.Labels{"file_sd_index": strconv.Itoa(i)},
```

Comment:
The index is an implementation detail, it shouldn't really be exposed to the user.

The challenge here really is how to deal with files that appear in multiple distinct scrape_configs.

Reply:
shouldn't the same file being in multiple file_sd or scrape_config jobs be invalid? but I guess in either case the solution is to have a single map and dedup entry's when we update the map

Reply:
It's perfectly valid to do that.

Reply:
so just to confirm, something like this is valid and needs to be handled properly by the timestamp collection code:
```
scrape_configs:
- job_name: "overwritten-default"
  scrape_interval: 10s

  file_sd_configs:
  - files:
    - filesd1.yml
    - filesd2.yml
  - files:
    - filesd2.yml
    - filesd3.yml
- job_name: "jobTwo"
  scrape_interval: 10s
  
  file_sd_configs:
    - filesd1.yml
    - filesd3.yml
```

Reply:
Yes